### DISCUSSION TITLE
How to open devtools and keep the focus on console as default

### DISCUSSION THREAD

**Discussion by hiyamandal** [seleniumbase/SeleniumBase#2066] [Q&A] [answered]
Hello, 

For a test, I need the devtools to be always open and view the console tab. At the end of the test, I need to print the logs. 
I tried the following code snippet but no luck. 

```
if devtools:
            self.driver.execute_cdp_cmd(
                "Console.enable", {}
            )

        self.open("https://example.com")
        # Perform your test actions

        if devtools:
            console_logs = self.driver.execute_cdp_cmd(
                "Console.getConsoleMessages", {}
            )
            for log in console_logs["messages"]:
                print(f"Console Log: {log['message']}")
```

Can anyone please help me how to 
1. Open the console tab
2. record and print the logs on console?

**Comment by mdmintz** [comment, marked as answer]
Hello,

For 1: Add ``--devtools`` as a pytest command-line option to open devtools with the browser. [More info here](https://github.com/seleniumbase/SeleniumBase/blob/63d95c42fc84bbcea415c6d8a3a201587b89c92e/help_docs/customizing_test_runs.md?plain=1#L185).

For 2: There are methods for that: (See the example test: [SeleniumBase/examples/test_console_logging.py](https://github.com/seleniumbase/SeleniumBase/blob/master/examples/test_console_logging.py))

```python
self.start_recording_console_logs()
"""
Starts recording console logs. Logs are saved to: "console.logs".
To get those logs later, call "self.get_recorded_console_logs()".
If navigating to a new page, then the current recorded logs will be
lost, and you'll have to call start_recording_console_logs() again.
# Link1: https://stackoverflow.com/a/19846113/7058266
# Link2: https://stackoverflow.com/a/74196986/7058266
"""

self.console_log_string(string)
"""
Log a string to the Web Browser's Console.
Example:
self.console_log_string("Hello World!")
"""

self.console_log_script(script)
"""
Log output of JavaScript to the Web Browser's Console.
Example:
self.console_log_script('document.querySelector("h2").textContent')
"""

self.get_recorded_console_logs()
"""
Returns console logs recorded after "start_recording_console_logs()".
"""
```

**Reply by hiyamandal** [reply]
Hi! 
I am using --devtools but the Element tab is in focus. I need to keep the console tab in focus. 

I tried using the console log recording APIs, but it is returning an empty dictionary. I am not navigating to a new page. For your reference, here's the code: 
<img width="837" alt="image" src="https://github.com/seleniumbase/SeleniumBase/assets/15275310/c1ded925-baf3-4272-9426-8b3316155bfd">

**Reply by mdmintz** [reply]
Try from the example and work from there: https://github.com/seleniumbase/SeleniumBase/blob/master/examples/test_console_logging.py
You can try Stack Overflow to see if you can find a way to control the focus of the Element Tab vs the Console Tab. I'm not aware of a way to specify that.

**Reply by hiyamandal** [reply]
I am sorry to just broadly say it is not working. The console logging works, but only records the errors. It is ignoring the warnings, etc. I tried the string and script logging API. Those APIs also working. 
Thanks for your time. I will keep posted if I can find out how to switch the tabs. I have seen some solutions using selenium but till now I have not succeeded in finding the working solution
